04_basic_struct/sysfs_list.c

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/kobject.h>
+#include <linux/list.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+#include <linux/sysfs.h>
+#include <linux/types.h>
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Victor Krasnoshchok <ct3da21164@protonmail.ch>");
+MODULE_DESCRIPTION("Interface to a linked list of strings through sysfs.");
+MODULE_VERSION("0.1");
+
+/* ------------------------------------------ */
+/* List-related logic and data manipulation.  */
+
+struct data_item {
+	struct list_head list_node;
+	char *payload;
+	size_t payload_len;
+};
+
+static struct list_head root_node = { NULL, NULL };
+
+static ssize_t add_node(struct list_head *parent, const char *node_data,
+			size_t node_data_len)
+{
+	struct data_item *new_item;
+	size_t payload_len;
+
+	if (parent == NULL || node_data == NULL)
+		return -EINVAL;
+
+	new_item = kmalloc(sizeof(struct data_item), GFP_KERNEL);
+	if (new_item == NULL)
+		return -ENOMEM;
+
+	payload_len = node_data_len + 1; /* Including \0 */
+	new_item->payload = kzalloc(payload_len, GFP_KERNEL);
+	if (new_item->payload == NULL) {
+		pr_err("Failed to alloc. %lu bytes for payload.", payload_len);
+		kfree(new_item);
+		return -ENOMEM;
+	}
+
+	strncpy(new_item->payload, node_data, node_data_len);
+	new_item->payload_len = node_data_len;
+
+	list_add_tail(&new_item->list_node, parent);
+
+	/* No extra newline needed here - echo appends NLs by itself */
+	pr_info("Node added: %s", new_item->payload);
+	return node_data_len;
+}
+
+static ssize_t serialize_list(struct list_head *list_root, char *output)
+{
+	ssize_t total_len = 0;
+	char *wr_ptr = output;
+	struct list_head *curr_node;
+	struct data_item *curr_data;
+
+	if (output == NULL)
+		return -EINVAL;
+
+	list_for_each(curr_node, list_root) {
+		curr_data = list_entry(curr_node, struct data_item, list_node);
+		strncpy(wr_ptr, curr_data->payload, curr_data->payload_len);
+		wr_ptr += curr_data->payload_len;
+		total_len += curr_data->payload_len;
+	}
+
+	return total_len;
+}
+
+static void destroy_list(struct list_head *list_root)
+{
+	struct data_item *curr_data_item, *placeholder;
+	size_t nodes_cnt = 0;
+
+	list_for_each_entry_safe(curr_data_item, placeholder, list_root,
+				  list_node) {
+		kfree(curr_data_item->payload);
+		list_del(&curr_data_item->list_node);
+		kfree(curr_data_item);
+		++nodes_cnt;
```

Comment:
It is not necessary please avoid pre-increment

Reply:
Sorry, but I didn't get what's wrong with pre-increment itself. As far as I remember, pre-incrementing works slightly faster (takes something like 1 or 2 instructions less than post-incrementing) by avoiding the need for a temporary storage for the value being incremented. Here I use nodes_cnt variable just to track how many nodes I had in the list before freeing it - it wasn't required in the assignment, I just added this log line for myself for debugging purposes, since I failed to find any API which returns list nodes count by a single call.

Reply:
Also I've just checked https://www.kernel.org/doc/html/v4.10/process/coding-style.html - there seems to be no recommendations regarding using or not using prefix increment operation. The only mentions of "increment" word are these:
> no space before the postfix increment & decrement unary operators:
> ++  --
> no space after the prefix increment & decrement unary operators:
> ++  --